File: apps/web/lib/api/analytics.ts
import { apiRequest, ApiClientError } from '@/lib/api/client';
import { API_BASE_URL } from '@/lib/api/endpoints';

export interface AnalyticsMetric {
  readonly id: string;
  readonly label: string;
  readonly value: number | null;
  readonly unit?: string;
  readonly change?: number | null;
  readonly severity?: 'neutral' | 'positive' | 'warning' | 'critical';
  readonly drilldown_count?: number;
}

export interface AnalyticsSeriesPoint {
  readonly label: string;
  readonly value: number;
  readonly series?: string;
}

export interface AnalyticsRecord {
  readonly id: string;
  readonly [key: string]: unknown;
}

export interface AnalyticsSection {
  readonly id: string;
  readonly title: string;
  readonly description?: string;
  readonly metrics?: readonly AnalyticsMetric[];
  readonly series?: readonly AnalyticsSeriesPoint[];
  readonly records?: readonly AnalyticsRecord[];
  readonly columns?: readonly string[];
}

export interface AnalyticsFilterOptions {
  readonly workspaces?: readonly string[];
  readonly project_types?: readonly string[];
  readonly providers?: readonly string[];
  readonly stacks?: readonly string[];
  readonly statuses?: readonly string[];
  readonly modules?: readonly string[];
  readonly severities?: readonly string[];
  readonly agents?: readonly string[];
  readonly languages?: readonly string[];
  readonly frameworks?: readonly string[];
}

export interface AnalyticsOverviewResponse {
  readonly generated_at: string;
  readonly period_start?: string;
  readonly period_end?: string;
  readonly metrics: readonly AnalyticsMetric[];
  readonly sections?: readonly AnalyticsSection[];
  readonly filters?: AnalyticsFilterOptions;
}

export interface AnalyticsFilters {
  readonly period: string;
  readonly workspace: string;
  readonly project_type: string;
  readonly provider: string;
  readonly stack: string;
  readonly status: string;
  readonly module: string;
  readonly severity: string;
  readonly agent: string;
  readonly language: string;
  readonly framework: string;
}

const SENSITIVE_KEY = /(?:api[_-]?key|access[_-]?token|refresh[_-]?token|secret|password|credential|authorization|prompt|raw[_-]?log)/i;

export function redactAnalyticsPayload(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(redactAnalyticsPayload);
  if (!value || typeof value !== 'object') return value;

  return Object.fromEntries(
    Object.entries(value)
      .filter(([key]) => !SENSITIVE_KEY.test(key))
      .map(([key, child]) => [key, redactAnalyticsPayload(child)]),
  );
}

function buildAnalyticsQuery(filters: AnalyticsFilters): string {
  const params = new URLSearchParams();
  Object.entries(filters).forEach(([key, value]) => {
    if (value && value !== 'all') params.set(key, value);
  });
  const query = params.toString();
  return query ? `?${query}` : '';
}

export async function getAnalyticsOverview(filters: AnalyticsFilters): Promise<AnalyticsOverviewResponse | null> {
  try {
    const payload = await apiRequest<unknown>(
      `${API_BASE_URL}/api/analytics/overview${buildAnalyticsQuery(filters)}`,
    );
    return redactAnalyticsPayload(payload) as AnalyticsOverviewResponse;
  } catch (error) {
    if (error instanceof ApiClientError && error.status === 404) return null;
    throw error;
  }
}

File: apps/api/app/schemas/analytics.py
"""Analytics Center V1 response contracts.

Field names mirror the existing frontend types in `apps/web/lib/api/analytics.ts`
(`AnalyticsMetric`, `AnalyticsSection`, `AnalyticsFilterOptions`,
`AnalyticsOverviewResponse`) so the Analytics Center page consumes this payload
without any adapter. Extra, non-sensitive fields (`source`, section `status` /
`reason`) are additive and safely ignored by the frontend.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

MetricSeverity = Literal["neutral", "positive", "warning", "critical"]
SectionStatus = Literal["available", "empty", "error"]


class AnalyticsMetric(BaseModel):
    id: str
    label: str
    value: float | int | None = None
    unit: str | None = None
    # Frontend reads `change`; kept nullable because V1 has no historical deltas.
    change: float | None = None
    severity: MetricSeverity = "neutral"
    drilldown_count: int | None = None
    # Which real module produced this metric (operational provenance).
    source: str | None = None


class AnalyticsSeriesPoint(BaseModel):
    label: str
    value: float | int
    series: str | None = None


class AnalyticsSection(BaseModel):
    id: str
    title: str
    description: str | None = None
    # `available` with data, `empty` with a reason, or `error` if the collector
    # failed in isolation (the rest of the response still succeeds).
    status: SectionStatus = "available"
    reason: str | None = None
    metrics: list[AnalyticsMetric] = Field(default_factory=list)
    series: list[AnalyticsSeriesPoint] = Field(default_factory=list)
    # Drill-down rows. Arbitrary, already-redacted, JSON-safe dicts; each has an `id`.
    records: list[dict[str, Any]] = Field(default_factory=list)
    columns: list[str] = Field(default_factory=list)


class AnalyticsFilterOptions(BaseModel):
    workspaces: list[str] = Field(default_factory=list)
    project_types: list[str] = Field(default_factory=list)
    providers: list[str] = Field(default_factory=list)
    stacks: list[str] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)
    modules: list[str] = Field(default_factory=list)
    severities: list[str] = Field(default_factory=list)
    agents: list[str] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    frameworks: list[str] = Field(default_factory=list)


class AnalyticsOverviewResponse(BaseModel):
    generated_at: str
    period_start: str | None = None
    period_end: str | None = None
    metrics: list[AnalyticsMetric] = Field(default_factory=list)
    sections: list[AnalyticsSection] = Field(default_factory=list)
    filters: AnalyticsFilterOptions = Field(default_factory=AnalyticsFilterOptions)
